lib/tree.rb

```diff
@@ -16,42 +16,144 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  # Time Complexity: O(n)
+  # Space Complexity: O(n)
   def add(key, value)
-    raise NotImplementedError
+    new_node = TreeNode.new(key,value)
+
+    if @root.nil?
+      @root = new_node
+    else
+      add_node(@root, new_node)
+    end
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_node(current, new_node)
+    return new_node if current.nil?
+
+    if new_node.key <= current.key
+      current.left = add_node(current.left, new_node)
+    else
+      current.right = add_node(current.right, new_node)
+    end
+
+    return current
+  end
+
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def find(key)
-    raise NotImplementedError
+    current = @root
+    return nil if current.nil?
+    
+    until current == nil
+      if current.key == key
+        return current.value
+      elsif key < current.key
+        current = current.left
+      else
+        current = current.right
+      end
+    end
+    return nil
   end
 
+  #used the following as inspiration for the traversals: http://rubyalgorithms.com/binary_search_tree.html
+
   # Time Complexity: 
   # Space Complexity: 
   def inorder
-    raise NotImplementedError
+    node = @root
+    traverse_array = []
+    return traverse_array if node.nil?
+    
+    in_order_helper(node) do |node|
+      traverse_array << {:key=> node.key, :value=>node.value}
+      # puts node.key
+    end
+
+    return traverse_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def in_order_helper(node, &block)
+    return if node.nil?
+    in_order_helper(node.left, &block)
+    yield node
+    in_order_helper(node.right, &block)
+  end
+
+  # Time Complexity: o(n)
+  # Space Complexity: O(n^2) 
   def preorder
-    raise NotImplementedError
+    node = @root
+    traverse_array = []
+    return traverse_array if node.nil?
+    
+    pre_order_helper(node) do |node|
+      traverse_array << {:key=> node.key, :value=>node.value}
+      # puts node.key
+    end
+
+    return traverse_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def pre_order_helper(node, &block)
+    return if node.nil?
+
+    yield node
+    pre_order_helper(node.left, &block)
+    pre_order_helper(node.right, &block)
+  end
+
+ # Time Complexity: o(n)
+  # Space Complexity: O(n^2) 
   def postorder
-    raise NotImplementedError
+    node = @root
+    traverse_array = []
+    return traverse_array if node.nil?
+    
+    post_order_helper(node) do |node|
+      traverse_array << {:key=> node.key, :value=>node.value}
+      # puts node.key
+    end
+
+    return traverse_array
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def post_order_helper(node, &block)
+    return if node.nil?
+
+    post_order_helper(node.left, &block)
+    post_order_helper(node.right, &block)
+    yield node
+  end
+
+  # Time Complexity: o(n)
+  # Space Complexity: O(n^2) 
   def height
```

Comment:
This doesn't actually work, as it only checks the left and right edges of the tree.  Instead you should do something like this:

For example, you wouldn't catch:

     5
   /  \
  3   10
 /      \
1        15
          /  \
        /    25
     14
   /
12

This does catch it.

```ruby
  def height_helper(current_node)
    return 0 if current_node.nil?
    return 1 + max(height_helper(current_node.left), height_helper(current_node.right))
  end
```